### PR TITLE
Add explicit reboot mode options

### PR DESCRIPTION
**Pull request by rdoxenham**
Currently the reboot-interface is not suitable for all circumstances due to the softPowerOff being enacted first, leading to a delay in the node being rebooted. The suggestion here is to make this more flexible and allow the operator to select whether they want to default to hard reboots instead.

### CODE REVIEW

File: design/baremetal-operator/reboot-interface.md
Comment (rdoxenham) [reply design/baremetal-operator/reboot-interface.md:176]: Thanks for looking into this @dhellmann.

I definitely agree that there's a need to maintain the softPowerOff() function for the reasons that you outlined, and hence the default outcome for setting a node to be rebooted should absolutely result in a graceful shutdown. However, there needs to be a way of forcing the reboot interface to go straight to hardPowerOff() if the user wants to do this.

I think that using the existing annotating and adapting it is likely the least disruptive - having the option to force hard, e.g. with `reboot.metal3.io = {'mode': 'hard'}` works well, and also adheres to the "default is soft" idea, but I'm not sure where you'd be able to specify this or how it would get looked up - I'm not familiar enough with how it's all wired together. Seems like the general time that you'd want the hard power off to be invoked would be in the event of the machine-health-check reporting unhealthy.

Comment (dhellmann) [reply design/baremetal-operator/reboot-interface.md:176]: Yes, any consumers of the API would need to be updated to add the extra value if they want hard shutdown.

It's not clear who is likely to be doing the work on this. I won't have time myself, so I was trying to provide some guidance about the first step (updating the API definition) to make it easier for someone else to pick up the implementation. Are you likely to have time to shepherd this through and maybe even implement it, @rdoxenham?

Comment (hardys) [reply design/baremetal-operator/reboot-interface.md:176]: Yes I'd reword this avoiding references to MachineHealthCheck etc, something like 

"Some users of the `reboot.metal3.io` API (and particularly fencing/remediation use-cases) require hard-reboot behavior to avoid any potential delay in the case where a soft-poweroff is unsuccessful" 

It's unfortunate that the `reboot.metal3.io/{key}` form appears to by-design accept any arbitrary key the design doc says `Callers indicate to the controller that they are ready to continue by removing the annotation with their unique {key} suffix.` - so I assume there was some use-case in mind for allowing that but I'm not sure what it was - @n1r1 perhaps you know?

It's also unfortunate that the removal of the `{key}` suffix appears to also include the case where a user specifies ``reboot.metal3.io: something`  https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/baremetalhost_controller.go#L737

So unless we redefine that existing behavior (which, arguably, is pretty low risk), we probably have two options:

1. New annotation as mentioned by @dhellmann e.g `reboot-mode.metal3.io/{soft,hard}`
2. Consider the current behavior a bug, and make `reboot.metal3.io` always hard-reboot (with the option of adding `soft-reboot.metal3.io` in future)

(2) has the advantage of not impacting any users that expect the original behavior of `reboot.metal3.io` - but clearly there's also the risk that users exist which now rely on the new soft-reboot behavior (I'm not aware of any, and since we've at different times done both, I guess we just have to pick one and ensure it's documented).
https://github.com/metal3-io/baremetal-operator/pull/424

Since the behavior of the `reboot.metal3.io` API changed after it landed in

Comment (n1r1) [reply design/baremetal-operator/reboot-interface.md:176]: > It's unfortunate that the `reboot.metal3.io/{key}` form appears to by-design accept any arbitrary key the design doc says `Callers indicate to the controller that they are ready to continue by removing the annotation with their unique {key} suffix.` - so I assume there was some use-case in mind for allowing that but I'm not sure what it was - @n1r1 perhaps you know?

The suffixed annotation (where `{key}` exists) is being used when the caller wants to keep the power off for a while (as opposed to immediate power-off and power-on). This allows the caller to do some stuff when the host is powered off.
I believe the idea was that multiple consumers might use the reboot api at the same time. 
In order to signal that the caller finished what it needs to do while the host was powered off, it should remove the annotation.
If there are two callers, using the same annotation, one caller might remove the annotation before the other caller finished his work on the powered off host.
By letting the caller use a unique key we ensure that callers will only remove their own annotation.
So if two callers added two different keys, BMO will power on the host only after both annotations removed.

The primary use-case for the suffixed annotation is fencing, where we need to delete the `node` from the cluster while the host is powered off .

Comment (n1r1) [reply design/baremetal-operator/reboot-interface.md:176]: > It's also unfortunate that the removal of the {key} suffix appears to also include the case where a user specifies reboot.metal3.io: something https://github.com/metal3-io/baremetal-operator/blob/master/controllers/metal3.io/baremetalhost_controller.go#L737

Not sure I understand this one. Can you please elaborate?

Comment (hardys) [reply design/baremetal-operator/reboot-interface.md:176]: Actually to deal with the multiple-actors problem mentioned, (1) could instead be `reboot-mode.metal3.io: force` or something, then it's an explicit key:value not an annotation which could potentially be specified in multiple conflicting ways?

Alternatively like (2) we could default to a hard-reboot for `reboot.metal3.io` and add a `reboot-mode.metal3.io: soft` if anyone turns out to need that behavior

Comment (andfasano) [reply design/baremetal-operator/reboot-interface.md:176]: Option #1 seems better to me, as it does not impact the existing behavior. Regarding the potential racing condition, what about keeping the existing pattern also for the annotation, ie `reboot-mode.metal3.io/{soft, hard}/{key}` ? This will allow BMO to detect/manage any eventual conflict

Comment (n1r1) [reply design/baremetal-operator/reboot-interface.md:176]: two thoughts:

1. Do we want to allow a soft shutdown request without escalating to hard shutdown?
2. We need to make sure we allow soft/hard shutdown for both cases - phased and immediate reboot. So practically there are 4 cases (more if want #1) - soft + immediate, soft + phased, hard + immediate, hard + phased.

Comment (n1r1) [reply design/baremetal-operator/reboot-interface.md:176]: > Actually to deal with the multiple-actors problem mentioned, (1) could instead be `reboot-mode.metal3.io: force` or something, then it's an explicit key:value not an annotation which could potentially be specified in multiple conflicting ways?
> 

In such case, I'd expect any controller who wants a soft shutdown, to remove the `reboot-mode.metal3.io: force`  annotation so we still have a race.
Also, this annotation might be added after some controller requested a soft shutdown.
Maybe we should first decide what's the required behavior before how to implement?
I'm not sure what we want to happen if one consumer requested soft and the other hard.

Comment (rdoxenham) [reply design/baremetal-operator/reboot-interface.md:176]: I think the soft shutdown request without escalating to hard shutdown is likely a case that we'd need to explore, but what alternatives do we have in that scenario? Do we assume that in the case of a soft shutdown that the reboot flag is only applied when, for example, software updates are complete and in theory it should be safe to forcefully reboot the node after a soft reboot fails?

> > Actually to deal with the multiple-actors problem mentioned, (1) could instead be `reboot-mode.metal3.io: force` or something, then it's an explicit key:value not an annotation which could potentially be specified in multiple conflicting ways?
> 
> In such case, I'd expect any controller who wants a soft shutdown, to remove the `reboot-mode.metal3.io: force` annotation so we still have a race.
> Also, this annotation might be added after some controller requested a soft shutdown.
> Maybe we should first decide what's the required behavior before how to implement?
> I'm not sure what we want to happen if one consumer requested soft and the other hard.

In my opinion, the workloads are paramount here. If one consumer requests soft, but another hard, the hard should always take precedent. A node can be fixed/rebuilt, but customers will likely expect their workload to be recovered (where possible) as quickly as possible on another node.

Comment (hardys) [reply design/baremetal-operator/reboot-interface.md:176]: > Not sure I understand this one. Can you please elaborate?

Sure, sorry I see now my comment wasn't clear - I was thinking about a more single-actor implementation of the `{key}` API, e.g:

```
reboot.metal3.io    # Current behavior, soft-reboot
reboot.metal3.io: force  # Hard reboot
reboot.metal3.io: wait   # wait before power-on
```

However I now the current API is designed to do e.g:

```
reboot.metal3.io/wait1
reboot.metal3.io/wait2
reboot.metal3.io
```

Having multiple actors triggering reboot sounds like a recipe for chaos to me, do such use-cases definitely exist, before we optimize the API for it?


>    Do we want to allow a soft shutdown request without escalating to hard shutdown?

I'm not aware of any requirement for that currently, but sure we should ensure any API changes don't preclude such an option in future.

> We need to make sure we allow soft/hard shutdown for both cases - phased and immediate reboot. So practically there are 4 cases (more if want #1) - soft + immediate, soft + phased, hard + immediate, hard + phased.

I'm not sure I follow the 4, perhaps you can clarify the difference between soft/phased, as I think there are 3?

* Soft reboot (escalate to hard reboot on failure)  # Behavior since https://github.com/metal3-io/baremetal-operator/pull/294
* Hard reboot (do not attempt soft-poweroff) # Behavior prior to https://github.com/metal3-io/baremetal-operator/pull/294
* Soft reboot (do nothing on failure)   # As above, we have no current requirement for this?

Comment (n1r1) [reply design/baremetal-operator/reboot-interface.md:176]: @hardys thanks for clarifying

> Having multiple actors triggering reboot sounds like a recipe for chaos to me, do such use-cases definitely exist, before we optimize the API for it?

I'm not aware of specific use-case for that, but since this is possible we should at least not violate the API promises.
If an actor added a suffixed annotation, we must ensure the host is powered off until that actor removes that annotation, regardless to what other actors are doing. 
Violating this, in the fencing use case, would mean a possible data corruption and violation of singletone applications (I can elaborate more if needed)

> > Do we want to allow a soft shutdown request without escalating to hard shutdown?
> 
> I'm not aware of any requirement for that currently, but sure we should ensure any API changes don't preclude such an option in future.

Agreed.

> > We need to make sure we allow soft/hard shutdown for both cases - phased and immediate reboot. So practically there are 4 cases (more if want #1) - soft + immediate, soft + phased, hard + immediate, hard + phased.
> 
> I'm not sure I follow the 4, perhaps you can clarify the difference between soft/phased, as I think there are 3?

The reboot api supports two types of annotations which translate to two APIs:

1. **suffixless annotation** - `reboot.metal3.io` - triggers full and immediate power cycle - BMO powers off the host and then immediately powers it back on. I called this API - "immediate reboot"
2.  **suffixed annotation** - `reboot.metal3.io/{suffix}` - triggers power off. BMO won't power the host back on until that annotation was removed by the one who added it. I called this API - "phased reboot"

Each on of these API should be able to support both soft and hard shutdown.
So we need to make sure we support all 4 possibilities:

- immediate reboot which is done by soft shutdown
- immediate reboot which is done by hard shutdown
- phased reboot which is done by soft shutdown
- phased reboot which is done by hard shutdown

If we want to add the "only soft shutdown" mode (i.e. do nothing on failure) we will have 2^3 options to support.

Comment (dhellmann) [reply design/baremetal-operator/reboot-interface.md:176]: Let's stick with always rebooting, and doing the best we can to do a soft reboot if that's requested.

Does my proposal for using a value with the annotation from earlier in this thread make sense as an API?

Comment (n1r1) [reply design/baremetal-operator/reboot-interface.md:176]: > Does my proposal for using a value with the annotation from earlier in this thread make sense as an API?

IMO yes.

> If there is no value for mode, we would assume the mode is soft.

I would just replace "there is no value for mode" with "the value does not equal to `{'mode': 'hard'}`", to handle the case where there's invalid value which may have been implemented by someone already

Comment (dhellmann) [reply design/baremetal-operator/reboot-interface.md:176]: > > Does my proposal for using a value with the annotation from earlier in this thread make sense as an API?
> 
> IMO yes.
> 
> > If there is no value for mode, we would assume the mode is soft.
> 
> I would just replace "there is no value for mode" with "the value does not equal to `{'mode': 'hard'}`", to handle the case where there's invalid value which may have been implemented by someone already

Good point, yes.

Does someone want to take a stab at updating the design doc with those new details?

Comment (hardys) [reply design/baremetal-operator/reboot-interface.md:176]: > Does someone want to take a stab at updating the design doc with those new details?

If @rdo doesn't have time I'm happy to, but first let me see if I can summarize exactly what the API would look like combining the feedback above:

```
reboot.metal3.io    # immediate reboot which is done by soft shutdown
reboot.metal3.io: {'mode': 'hard'}    # immediate reboot which is done by hard shutdown
reboot.metal3.io/{key}    # phased reboot which is done by soft shutdown
reboot.metal3.io/{key}: {'mode': 'hard'}    # phased reboot which is done by hard shutdown
```

Then if we decide in future that `soft shutdown without escalating to hard shutdown` is a thing we actually need, it could be some other mode (but we won't add that to the proposal now, since it's not clear anyone needs it, and in general we want to avoid getting stuck forever after a failed soft-poweroff)

Does that sound right?

Comment (dhellmann) [reply design/baremetal-operator/reboot-interface.md:176]: Yes, that's the idea.

We should probably also allow `{'mode': 'soft'}` for completeness, but those cases are the same as the first and third in your list.

Comment (rdoxenham) [reply design/baremetal-operator/reboot-interface.md:176]: Sure, I'll push another update to this PR. Thanks all!

Comment (hardys) [reply design/baremetal-operator/reboot-interface.md:176]: > We should probably also allow {'mode': 'soft'} for completeness, but those cases are the same as the first and third in your list.

Yup, my assumption was that we'd just ignore any unsupported modes (as mentioned by @n1r1 above) - so this case would just result in the same behavior as not specifying any mode.

File: design/baremetal-operator/reboot-interface.md
Comment (dhellmann) [inline design/baremetal-operator/reboot-interface.md:151]:
```suggestion
- `reboot.metal3.io/{key}` -- phased reboot, issued and managed by the client
```

File: design/baremetal-operator/reboot-interface.md
Comment (dhellmann) [inline design/baremetal-operator/reboot-interface.md:155]:
```suggestion
- `reboot.metal3.io:/{key}: {'mode':'hard'}` -- phased reboot, issued and
```

File: design/baremetal-operator/reboot-interface.md
Comment (dhellmann) [inline design/baremetal-operator/reboot-interface.md:145]:
```suggestion
- `reboot.metal.io` -- immediate reboot via soft shutdown first, followed by
```

File: design/baremetal-operator/reboot-interface.md
Comment (dhellmann) [inline design/baremetal-operator/reboot-interface.md:148]:
```suggestion
- `reboot.metal3.io: {'mode':'hard'}` -- immediate reboot via hard shutdown,
```

**Comment by dhellmann** [comment]
I think this accurately describes what we agreed in the comment thread.

/approve

File: design/baremetal-operator/reboot-interface.md
Comment (n1r1) [inline design/baremetal-operator/reboot-interface.md:152]: I think the colons are in wrong place?
should be  `reboot.metal3.io/{key}: {'mode':'hard'}`

File: design/baremetal-operator/reboot-interface.md
Comment (n1r1) [inline design/baremetal-operator/reboot-interface.md:156]: The way I read it is that `{'mode':'foo'}` will trigger soft reboot, but there's no description for any other value, such as just `foo`.
There are already clients who are using this annotation with some value (e.g. OCP UI), so maybe I'd replace `mode` with `value`.

**Review by n1r1 — COMMENTED** [review]
Looks good.
Left two small comments, otherwise lgtm

File: design/baremetal-operator/reboot-interface.md
Comment (rdoxenham) [reply design/baremetal-operator/reboot-interface.md:152]: Good catch, thank you!

File: design/baremetal-operator/reboot-interface.md
Comment (hardys) [reply design/baremetal-operator/reboot-interface.md:156]: The annotation, and values/schema supported should be owned by the BMH/BMO - it doesn't seem right that external systems are (ab)using the annotation to store state - how does changing the key from `mode` to `value` help here if we have no control over whatever arbitrary data may be written?

Comment (rdoxenham) [reply design/baremetal-operator/reboot-interface.md]: Thanks, have clarified.

File: design/baremetal-operator/reboot-interface.md
Comment (dhellmann) [inline design/baremetal-operator/reboot-interface.md:158]: This is mixing the default when we get an error with the description of the API.

The argument to the annotation isn't just a string or map, it's the representation of an object. The object has 1 member `mode` that can have 2 values of an enum, `hard` or `soft`. If we can't interpret the data given to us as fitting into that description, the default for `mode` is `soft`.

**Review by dhellmann — CHANGES_REQUESTED** [review]

File: design/baremetal-operator/reboot-interface.md
Comment (rdoxenham) [reply design/baremetal-operator/reboot-interface.md]: Thanks, I've addressed this in the latest commit. WDYT?

File: design/baremetal-operator/reboot-interface.md
Comment (dhellmann) [inline design/baremetal-operator/reboot-interface.md:156]: Then remove this paragraph. I don't think we need to emphasize the error handling, since that's implicit in describing what we expect to find in the map above.

File: design/baremetal-operator/reboot-interface.md
Comment (dhellmann) [inline design/baremetal-operator/reboot-interface.md:143]: How about if we change this to describe the expectations for the value

```suggestion
The annotation value should be a JSON map containing a key `'mode'` with values of `'hard'` or `'soft'` to specify the reboot mode. The default reboot mode is `'soft'`.

These are a few examples of using the reboot API
```

File: design/baremetal-operator/reboot-interface.md
Comment (rdoxenham) [reply design/baremetal-operator/reboot-interface.md:156]: Done

File: design/baremetal-operator/reboot-interface.md
Comment (rdoxenham) [reply design/baremetal-operator/reboot-interface.md:143]: Yep, that's fine. Changing.

**Comment by fmuyassarov** [comment]
Let's have a  week of lazy consensus, as there is no objections for this from the community.

**Comment by dhellmann** [comment]
/lgtm
/hold

I'm going to be out next week, and some of the other folks in the OWNERS file are also out. So, in the interest of not holding folks up, I'll set this up so someone can just remove the hold when it's ready to merge. @fmuyassarov, maybe you can do that?

**Comment by fmuyassarov** [comment]
> /lgtm
> /hold
> 
> I'm going to be out next week, and some of the other folks in the OWNERS file are also out. So, in the interest of not holding folks up, I'll set this up so someone can just remove the hold when it's ready to merge. @fmuyassarov, maybe you can do that?

Sure, if there are no more comments to be addresses by the next week, I will remove the hold.

**Comment by dhellmann** [comment]
> Sure, if there are no more comments to be addressed by the next week, I will remove the hold.

Thanks!

**Comment by fmuyassarov** [comment]
It seems there were no any further comments/objections to be addressed on this proposal. As agreed last week, today is the end of the lazy consensus. As such, I'm removing the hold. 
/hold cancel

**Pull request merged**
